include/nbl/builtin/hlsl/bda/__ref.hlsl

```diff
@@ -0,0 +1,128 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/cpp_compat.hlsl"
+#include "nbl/builtin/hlsl/spirv_intrinsics/core.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_BDA_REF_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BDA_REF_INCLUDED_
+
+namespace spirv
+{
+template<typename M, typename T, typename StorageClass>
+[[vk::ext_instruction(/*spv::OpAccessChain*/65)]]
+vk::SpirvOpaqueType </* OpTypePointer*/ 32,StorageClass,M> accessChain(
+    [[vk::ext_reference]] vk::SpirvOpaqueType </* OpTypePointer*/ 32,
+    StorageClass,T>base,
+    [[vk::ext_literal]] uint32_t index0
+);
+}
+
+namespace bda
+{
+template<typename T>
+using __spv_ptr_t = vk::SpirvOpaqueType<
+    /* OpTypePointer */ 32,
+    /* PhysicalStorageBuffer */ vk::Literal<vk::integral_constant<uint,5349> >,
+    T
+>;
+
+namespace impl
+{
+// this only exists to workaround DXC issue XYZW TODO https://github.com/microsoft/DirectXShaderCompiler/issues/6576
+template<class T>
+[[vk::ext_capability(/*PhysicalStorageBufferAddresses */ 5347 )]]
+[[vk::ext_instruction(/*spv::OpBitcast*/124)]]
+T bitcast(uint64_t);
+
+template<typename T, typename P, uint32_t alignment>
+[[vk::ext_capability( /*PhysicalStorageBufferAddresses */5347)]]
+[[vk::ext_instruction( /*OpLoad*/61)]]
+T load(P pointer, [[vk::ext_literal]] uint32_t __aligned = /*Aligned*/0x00000002, [[vk::ext_literal]] uint32_t __alignment = alignment);
+
+template<typename T, typename P, uint32_t alignment >
+[[vk::ext_capability( /*PhysicalStorageBufferAddresses */5347)]]
+[[vk::ext_instruction( /*OpStore*/62)]]
+void store(P pointer, T obj, [[vk::ext_literal]] uint32_t __aligned = /*Aligned*/0x00000002, [[vk::ext_literal]] uint32_t __alignment = alignment);
+
+// TODO: atomics for different types
+template<typename T, typename P> // integers operate on 2s complement so same op for signed and unsigned
+[[vk::ext_instruction( /*spv::OpAtomicIAdd*/234)]]
+T atomicIAdd(P ptr, uint32_t memoryScope, uint32_t memorySemantics, T value);
```

Comment:
we should already have those in `spirv_intrinsics` you might need to add overloads though, like `load` and `store` where the ones that take alignment specify an extra capability, or the bitcast

Reply:
I know that any atomic, needs two versions due to limitations of DXC
```cpp
template<typename T>
[[vk::ext_instruction(spv::OpAtomicLoad)]]
T atomicLoad([[vk::ext_reference]] T pointer, uint32_t memory, uint32_t semantics);

template<typename T, typename Ptr_T>
[[vk::ext_instruction(spv::OpAtomicLoad)]]
T atomicLoad(Ptr_T pointer, uint32_t memory, uint32_t semantics);
```

Reply:
Done